azure/services/natgateways/natgateways.go

```diff
@@ -55,119 +57,61 @@ func New(scope NatGatewayScope) *Service {
-
-		switch {
-		case err != nil && !azure.ResourceNotFound(err):
-			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
-		case err == nil:
-			// nat gateway already exists
-			s.Scope.V(4).Info("nat gateway already exists", "nat gateway", natGatewaySpec.Name)
-			natGatewaySpec.Subnet.NatGateway.ID = existingNatGateway.ID
-
-			if existingNatGateway.NatGatewayIP.Name == natGatewaySpec.NatGatewayIP.Name {
-				// Skip update for Nat Gateway as it exists with expected values
-				s.Scope.V(4).Info("Nat Gateway exists with expected values, skipping update", "nat gateway", natGatewaySpec.Name)
-				natGatewaySpec.Subnet.NatGateway = *existingNatGateway
-				s.Scope.SetSubnet(natGatewaySpec.Subnet)
-				continue
-			} else {
-				s.Scope.V(2).Info("updating NAT gateway IP name to match the spec", "old name", existingNatGateway.NatGatewayIP.Name, "desired name", natGatewaySpec.NatGatewayIP.Name)
-			}
-		default:
-			// nat gateway doesn't exist but its name was specified in the subnet, let's create it
-			s.Scope.V(2).Info("nat gateway doesn't exist yet, creating it", "nat gateway", natGatewaySpec.Name)
-		}
-
-		natGatewayToCreate := network.NatGateway{
-			Location: to.StringPtr(s.Scope.Location()),
-			Sku:      &network.NatGatewaySku{Name: network.Standard},
-			NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
-				PublicIPAddresses: &[]network.SubResource{
-					{
-						ID: to.StringPtr(azure.PublicIPID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.NatGatewayIP.Name)),
-					},
-				},
-			},
-		}
-		err = s.client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), natGatewaySpec.Name, natGatewayToCreate)
+		natGateway, err := async.CreateResource(ctx, s.Scope, s.client, natGatewaySpec, serviceName)
 		if err != nil {
-			return errors.Wrapf(err, "failed to create nat gateway %s in resource group %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+			if !azure.IsOperationNotDoneError(err) || result == nil {
+				result = err
+			}
 		}
-		s.Scope.V(2).Info("successfully created nat gateway", "nat gateway", natGatewaySpec.Name)
-		natGateway := infrav1.NatGateway{
-			ID:   azure.NatGatewayID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.Name),
-			Name: natGatewaySpec.Name,
-			NatGatewayIP: infrav1.PublicIPSpec{
-				Name: natGatewaySpec.NatGatewayIP.Name,
-			},
+		if err == nil && result != nil {
+			// TODO: consider making OwnerResourceName() a no-op
+			subnetSpec := s.Scope.Subnet(natGatewaySpec.OwnerResourceName())
+			networkNatGateway, ok := natGateway.(network.NatGateway)
+			if !ok {
+				// TODO: maybe this error shouldn't take priority
+				result = errors.Errorf("created resource %T is not a network.NatGateway", natGateway)
+			}
+			// TODO: is it necessary to set the spec since it doesn't appear to change any behavior
+			subnetSpec.NatGateway.ID = *networkNatGateway.ID
+			s.Scope.SetSubnet(subnetSpec)
```

Comment:
@CecileRobertMichon @mboersma Since the NAT gateway should already exist in the subnet spec, do we need to update the ID and set the subnet spec? I don't think that'll change the behavior if we remove it, and that way it'll be easy to make OwnerResourceName() a no-op instead of giving the subnet name.

Reply:
I agree with you it's not needed but in the spirit of not changing behavior as part of a refactor, let's keep setting the ID in case anyone was using it before / relying on it in some way.

You should still make `OwnerResourceName()` a no-op since the subnet isn't the owner resource of the NAT GW in this case.

What we could do here is add a func to the scope `SetNatGatewayID(name string,  ID string)` which ranges through the subnets and sets the ID for all the NAT gateways with a matching name to the spec. Note that one NAT GW may be used by more than one subnet so it's important to range through all of them, not just the first one we find (since we're not doing a 1:1 mapping anymore by passing in the subnet who asked for this NAT GW to be created).